devito/petsc/iet/passes.py

```diff
@@ -0,0 +1,366 @@
+from devito.passes.iet.engine import iet_pass
+from devito.ir.iet import (FindNodes, Call,
+                           Transformer, FindSymbols,
+                           MapNodes, Iteration, Callable, Callback, List, Uxreplace,
+                           Definition, BlankLine, PointerCast)
+from devito.petsc.types import (PetscMPIInt, PETScStruct, DM, Mat,
+                                Vec, KSP, PC, SNES, PetscErrorCode, PETScArray)
+from devito.symbolics import Byref, Macro, FieldFromPointer
+from devito.petsc.iet.nodes import MatVecAction, LinearSolverExpression
+from devito.petsc.utils import linear_solver_mapper
+import cgen as c
+
+
+@iet_pass
+def lower_petsc(iet, **kwargs):
+
+    # Check if PETScSolve was used.
+    petsc_nodes = FindNodes(MatVecAction).visit(iet)
+
+    if not petsc_nodes:
+        return iet, {}
+
+    else:
+        # Collect all petsc solution fields
+        unique_targets = list(set([i.expr.rhs.target for i in petsc_nodes]))
+
+        # Initalize PETSc
+        init = init_petsc(**kwargs)
+
+        # Create context data struct
+        struct = build_struct(iet)
+
+        objs = build_core_objects(unique_targets[-1], **kwargs)
+
+        # Create core PETSc calls (not specific to each PETScSolve)
+        core = core_petsc(unique_targets[-1], struct, objs, **kwargs)
+
+        matvec_mapper = MapNodes(Iteration, MatVecAction, 'groupby').visit(iet)
+
+        main_mapper = {}
+
+        setup = []
+        efuncs = []
+
+        for target in unique_targets:
+
+            solver_objs = build_solver_objs(target)
+
+            matvec_body_list = List()
+
+            solver_setup = False
```

Comment:
you may probably get rid of this boolean flag by looping twice over `for iter, (matvec,) in matvec_mapper.items():`, each time with a specific lowering objective